src/modules/ekf2/EKF/output_predictor.cpp

```diff
@@ -48,126 +48,183 @@ void OutputPredictor::print_status()
 	printf("output buffer: %d/%d (%d Bytes)\n", _output_buffer.entries(), _output_buffer.get_length(),
 	       _output_buffer.get_total_size());
-
-	printf("output vert buffer: %d/%d (%d Bytes)\n", _output_vert_buffer.entries(), _output_vert_buffer.get_length(),
-	       _output_vert_buffer.get_total_size());
-}
-
-void OutputPredictor::alignOutputFilter(const Quatf &quat_state, const Vector3f &vel_state, const Vector3f &pos_state)
-{
-	const outputSample &output_delayed = _output_buffer.get_oldest();
-
-	// calculate the quaternion rotation delta from the EKF to output observer states at the EKF fusion time horizon
-	Quatf q_delta{quat_state * output_delayed.quat_nominal.inversed()};
-	q_delta.normalize();
-
-	// calculate the velocity and position deltas between the output and EKF at the EKF fusion time horizon
-	const Vector3f vel_delta = vel_state - output_delayed.vel;
-	const Vector3f pos_delta = pos_state - output_delayed.pos;
-
-	// loop through the output filter state history and add the deltas
-	for (uint8_t i = 0; i < _output_buffer.get_length(); i++) {
-		_output_buffer[i].quat_nominal = q_delta * _output_buffer[i].quat_nominal;
-		_output_buffer[i].quat_nominal.normalize();
-		_output_buffer[i].vel += vel_delta;
-		_output_buffer[i].pos += pos_delta;
-	}
-
-	_output_new = _output_buffer.get_newest();
 }
 
 void OutputPredictor::reset()
 {
-	// TODO: who resets the output buffer content?
-	_output_new = {};
-	_output_vert_new = {};
+	_output_buffer.reset();
 
 	_accel_bias.setZero();
 	_gyro_bias.setZero();
 
 	_time_last_update_states_us = 0;
 	_time_last_correct_states_us = 0;
 
+	_output_new = {};
 	_R_to_earth_now.setIdentity();
-	_vel_imu_rel_body_ned.setZero();
-	_vel_deriv.setZero();
-
-	_delta_angle_corr.setZero();
+	_vel_imu_rel_body_ned.zero();
+	_vel_deriv.zero();
 
-	_vel_err_integ.setZero();
-	_pos_err_integ.setZero();
+	_delta_angle_corr.zero();
+	_vel_err_integ.zero();
+	_pos_err_integ.zero();
 
-	_output_tracking_error.setZero();
-
-	for (uint8_t index = 0; index < _output_buffer.get_length(); index++) {
-		_output_buffer[index] = {};
-	}
+	_output_tracking_error.zero();
 
-	for (uint8_t index = 0; index < _output_vert_buffer.get_length(); index++) {
-		_output_vert_buffer[index] = {};
-	}
+	_output_filter_aligned = false;
 }
 
-void OutputPredictor::resetQuaternion(const Quatf &quat_change)
+void OutputPredictor::resetQuaternion(const Quatf &new_quat)
 {
+	const outputSample &output_delayed = _output_buffer.get_oldest();
+	const Quatf quat_change{(new_quat * output_delayed.quat_nominal.inversed()).normalized()};
+
 	// add the reset amount to the output observer buffered data
 	for (uint8_t i = 0; i < _output_buffer.get_length(); i++) {
-		_output_buffer[i].quat_nominal = quat_change * _output_buffer[i].quat_nominal;
+		_output_buffer[i].quat_nominal = (quat_change * _output_buffer[i].quat_nominal).normalized();
+
+		// apply change to velocity
+		_output_buffer[i].vel = quat_change.rotateVector(_output_buffer[i].vel);
```

Comment:
If we need to rotate more than 1 vector, it gets more efficient to first convert to a rotation matrix.